MailCore/Cache/MailboxManager/MailboxManager+Message.swift

```diff
@@ -469,15 +478,51 @@ public extension MailboxManager {
         for thread in existingThreads {
             for existingMessage in allExistingMessages {
                 if !thread.messages.map(\.uid).contains(existingMessage.uid) {
-                    thread.addMessageIfNeeded(newMessage: message.fresh(using: realm) ?? message)
+                    thread.addMessageIfNeeded(newMessage: existingMessage.fresh(using: realm) ?? existingMessage)
                 }
             }
 
             threadsToUpdate.insert(thread)
         }
+
+        // Now that all other existing Threads are updated, we need to remove the duplicated Threads.
+        if isThereDuplicatedThreads {
+            removeDuplicatedThreads(messageIds: message.linkedUids, threadsToUpdate: &threadsToUpdate, using: realm)
+        }
+
         return threadsToUpdate
     }
 
+    private func isThereDuplicatedThreads(realm: Realm, messageIds: MutableSet<String>, threadCount: Int) -> Bool {
+        let foldersCount = realm.objects(Thread.self).where { $0.messageIds.containsAny(in: messageIds) }
+            .distinct(by: [\Thread.folderId]).count
+        return foldersCount != threadCount
+    }
+
+    private func removeDuplicatedThreads(
+        messageIds: MutableSet<String>,
+        threadsToUpdate: inout Set<Thread>,
+        using realm: Realm
+    ) {
+        // Create a map with all duplicated Threads of the same Thread in a list.
+        var map = [String: [Thread]]()
+        let threads = realm.objects(Thread.self).where { $0.messageIds.containsAny(in: messageIds) }
+        for thread in threads {
```

Comment:
I think that you could use `init(grouping:by:)`
https://developer.apple.com/documentation/swift/dictionary/init(grouping:by:)